src/autosuggest/autosuggest-input.tsx

```diff
@@ -0,0 +1,265 @@
+        inputRef.current?.focus();
+      },
+      open: openDropdown,
+      close: closeDropdown,
+    }));
+
+    const handleBlur: React.FocusEventHandler = event => {
+      const relatedTarget = getBlurEventRelatedTarget(event.nativeEvent);
+      if (event.currentTarget.contains(relatedTarget) || dropdownFooterRef.current?.contains(relatedTarget)) {
+        return;
+      }
+      closeDropdown();
+      fireNonCancelableEvent(onBlur, null);
+    };
+
+    const handleFocus = () => {
+      if (!preventOpenOnFocusRef.current) {
+        openDropdown();
+        fireNonCancelableEvent(onFocus, null);
+      }
+      preventOpenOnFocusRef.current = false;
+    };
+
+    const handleMouseDown: React.MouseEventHandler = event => {
+      // Prevent currently focused element from losing focus.
+      event.preventDefault();
+    };
+
+    const handleKeyDown = (e: CustomEvent<BaseKeyDetail>) => {
+      switch (e.detail.keyCode) {
+        case KeyCode.down: {
+          onPressArrowDown?.();
+          openDropdown();
+          e.preventDefault();
+          break;
+        }
+        case KeyCode.up: {
+          onPressArrowUp?.();
+          openDropdown();
+          e.preventDefault();
+          break;
+        }
+        case KeyCode.enter: {
+          if (open) {
+            if (!onPressEnter?.()) {
+              closeDropdown();
+            }
+            e.preventDefault();
+          }
+          fireCancelableEvent(onKeyDown, e.detail);
+          break;
+        }
+        case KeyCode.escape: {
+          if (open) {
+            closeDropdown();
+          } else if (value) {
+            fireNonCancelableEvent(onChange, { value: '' });
+          }
+          e.preventDefault();
+          fireCancelableEvent(onKeyDown, e.detail);
+          break;
+        }
+        default: {
+          fireCancelableEvent(onKeyDown, e.detail);
+        }
+      }
+    };
+
+    const handleChange = (value: string) => {
+      openDropdown();
+      fireNonCancelableEvent(onChange, { value });
+    };
+
+    const handleDelayedInput = (value: string) => {
+      fireNonCancelableEvent(onDelayedInput, { value });
+    };
+
+    const expanded = open && (dropdownExpanded ?? !!dropdownContent);
+    const nativeAttributes = {
+      name,
+      placeholder,
+      autoFocus,
+      onClick: openDropdown,
+      role: 'combobox',
+      'aria-autocomplete': 'list',
+      'aria-expanded': expanded,
+      'aria-controls': ariaControls,
+      // 'aria-owns' needed for safari+vo to announce activedescendant content
+      'aria-owns': ariaControls,
+      'aria-label': ariaLabel,
+      'aria-activedescendant': ariaActivedescendant,
+    };
+
+    const [trapDropdownFocus, setTrapDropdownFocus] = useState(false);
+
+    // Run this effect on every render to determine if necessary to trap focus around input and dropdown.
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    useEffect(() => {
+      setTrapDropdownFocus(
```

Comment:
So, the new React docs explicitly flag this as an anti-pattern: https://beta.reactjs.org/learn/you-might-not-need-an-effect#updating-state-based-on-props-or-state

Can we statically compute this value? Perhaps, introducing this as a prop and compute on the parent component level

Reply:
I tried to compute it w/o the state/effect but that does not trigger the dropdown to update. I will check again to see what can be done. It is possible to make it a prop but I'd like to avoid that.

Reply:
I am not sure the existing implementation is considered an anti-pattern according to the linked article. The reason of the useEffect is that we check the refs which only makes sense when the corresponding elements are rendered. If the computation is run inside the component's body, the refs would be null.

Reply:
If we have to disable the eslint plugin, this is a definite incompatibility with React idiomatic patterns.

If you believe this is not a risk, I will have to trust you on that.

Reply:
Thanks 👍